src/librbd/ImageCtx.cc

```diff
@@ -274,6 +274,11 @@ struct C_InvalidateCache : public Context {
     trace_endpoint.copy_name(pname);
     perf_start(pname);
 
+    assert(image_watcher == NULL);
+    image_watcher = new ImageWatcher<>(*this);
```

Comment:
Any reason why the ```ImageWatcher``` needs to be constructed prior to the refresh state machine? I grepped and didn't immediately see it being referenced (except for the change below to determine if it's safe to apply metadata string overrides).

Reply:
In refresh we create exclusive lock, and it (ManagedLock) references the watcher.

Reply:
Ack -- thanks